## HW 2 
## Hongbo Liu

1. Peaks 


a) What proportion of peaks is still unclimbed?

In [3]:
import pandas as pd
import numpy as np

In [4]:
peaks=pd.read_csv("C:/Users/22151/Documents/GitHub/course_content_5072/course_content/Exercises/HW02/data/peaks.csv")

In [5]:
peaks.head()

peak_id      peak_name peak_alternative_name  height_metres climbing_status  \
0    AMAD     Ama Dablam          Amai Dablang           6814         Climbed   
1    AMPG  Amphu Gyabjen                   NaN           5630         Climbed   
2    ANN1    Annapurna I                   NaN           8091         Climbed   
3    ANN2   Annapurna II                   NaN           7937         Climbed   
4    ANN3  Annapurna III                   NaN           7555         Climbed   

   first_ascent_year  first_ascent_country first_ascent_expedition_id  
0             1961.0  New Zealand, USA, UK                  AMAD61101  
1             1953.0                    UK                  AMPG53101  
2             1950.0                France                  ANN150101  
3             1960.0             UK, Nepal                  ANN260101  
4             1961.0                 India                  ANN361101

In [6]:
k=peaks.loc[peaks["climbing_status"]=="Climbed"]
r_1=len(k)
r_2=len(peaks)
prop=(r_2-r_1)/r_2
prop

0.27136752136752135

In [7]:
print("27.1% of all peaks are still unclimbed ")

27.1% of all peaks are still unclimbed 


b) What is the average height of the climbed peaks vs. unclimbed peaks?

In [8]:
k=peaks.loc[peaks["climbing_status"]=="Climbed"]
kk=peaks.loc[peaks["climbing_status"]!="Climbed"]
h_c=k["height_metres"].mean()
h_uc=kk["height_metres"].mean()

In [9]:
peaks.groupby("climbing_status")["height_metres"].mean()

climbing_status
Climbed      6706.284457
Unclimbed    6523.330709
Name: height_metres, dtype: float64

In [10]:
print("The average height of the climbed peaks is {}".format(h_c))

The average height of the climbed peaks is 6706.284457478006


In [11]:
print("The average height of the unclimbed peaks is {}".format(h_uc))

The average height of the unclimbed peaks is 6523.330708661418


2. Sherpas

a) What proportion of them was hired for the expedition? 

In [12]:
members=pd.read_csv("C:/Users/22151/Documents/GitHub/course_content_5072/course_content/Exercises/HW02/data/members.csv")

In [13]:
members

expedition_id     member_id peak_id        peak_name  year  season sex  \
0         AMAD78301  AMAD78301-01    AMAD       Ama Dablam  1978  Autumn   M   
1         AMAD78301  AMAD78301-02    AMAD       Ama Dablam  1978  Autumn   M   
2         AMAD78301  AMAD78301-03    AMAD       Ama Dablam  1978  Autumn   M   
3         AMAD78301  AMAD78301-04    AMAD       Ama Dablam  1978  Autumn   M   
4         AMAD78301  AMAD78301-05    AMAD       Ama Dablam  1978  Autumn   M   
...             ...           ...     ...              ...   ...     ...  ..   
76514     PUTH19101  PUTH19101-07    PUTH  Putha Hiunchuli  2019  Spring   M   
76515     RATC19101  RATC19101-01    RATC      Ratna Chuli  2019  Spring   M   
76516     SANK19101  SANK19101-01    SANK     Sano Kailash  2019  Spring   M   
76517     SANK19101  SANK19101-02    SANK     Sano Kailash  2019  Spring   F   
76518     SANK19101  SANK19101-03    SANK     Sano Kailash  2019  Spring   M   

        age citizenship expedition_role  ...  highpoint_metres  success  \
0      40.0      France          Leader  ...               NaN    False   
1      41.0      France   Deputy Leader  ...            6000.0    False   
2      27.0      France         Climber  ...               NaN    False   
3      40.0      France      Exp Doctor  ...            6000.0    False   
4      34.0      France         Climber  ...               NaN    False   
...     ...         ...             ...  ...               ...      ...   
76514  44.0       Nepal      H-A Worker  ...            6350.0    False   
76515  55.0     Romania          Leader  ...            6600.0    False   
76516  56.0     Austria          Leader  ...            6452.0     True   
76517  57.0     Austria         Climber  ...            5700.0    False   
76518  41.0       Nepal      H-A Worker  ...            6452.0     True   

        solo  oxygen_used   died  death_cause death_height_metres  injured  \
0      False        False  False          NaN                 NaN    False   
1      False        False  False          NaN                 NaN    False   
2      False        False  False          NaN                 NaN    False   
3      False        False  False          NaN                 NaN    False   
4      False        False  False          NaN                 NaN    False   
...      ...          ...    ...          ...                 ...      ...   
76514  False        False  False          NaN                 NaN    False   
76515  False        False  False          NaN                 NaN    False   
76516  False        False  False          NaN                 NaN    False   
76517  False        False  False          NaN                 NaN    False   
76518  False        False  False          NaN                 NaN    False   

       injury_type injury_height_metres  
0              NaN                  NaN  
1              NaN                  NaN  
2              NaN                  NaN  
3              NaN                  NaN  
4              NaN                  NaN  
...            ...                  ...  
76514          NaN                  NaN  
76515          NaN                  NaN  
76516          NaN                  NaN  
76517          NaN                  NaN  
76518          NaN                  NaN  

[76519 rows x 21 columns]

In [14]:
members["hired"].head()

0    False
1    False
2    False
3    False
4    False
Name: hired, dtype: bool

In [15]:
mem=members.loc[members["citizenship"]=="Nepal"]

In [16]:
m_e=mem.loc[mem["hired"]==True]
h_e=len(m_e)
t_e=len(mem)
prop_h=(t_e-h_e)/t_e
prop_h

0.06594360086767896

In [17]:
print("6.6% of the members were hired for the expedition")

6.6% of the members were hired for the expedition


b) What are their minimum, maximum, and average ages?

In [18]:
mem[["hired","age"]].groupby("hired").agg(['min', 'max', 'mean'])

age                 
        min   max       mean
hired                       
False  12.0  85.0  32.538542
True   15.0  67.0  32.909864

The minimum age for members who were hired for the expedition is 15, the maximum age is 67 and the average age is 32.9

The minimum age for members who were not hired for the expedition is 12, the maximum age is 85 and the average age is 32.53

3. Gender

In [93]:
expeditions=pd.read_csv("C:/Users/22151/Documents/GitHub/course_content_5072/course_content/Exercises/HW02/data/expeditions.csv")

In [20]:
expeditions

expedition_id peak_id        peak_name  year  season basecamp_date  \
0         ANN260101    ANN2     Annapurna II  1960  Spring    1960-03-15   
1         ANN269301    ANN2     Annapurna II  1969  Autumn    1969-09-25   
2         ANN273101    ANN2     Annapurna II  1973  Spring    1973-03-16   
3         ANN278301    ANN2     Annapurna II  1978  Autumn    1978-09-08   
4         ANN279301    ANN2     Annapurna II  1979  Autumn           NaN   
...             ...     ...              ...   ...     ...           ...   
10359     PUMO19101    PUMO           Pumori  2019  Spring    2019-04-04   
10360     PUMO19102    PUMO           Pumori  2019  Spring    2019-04-14   
10361     PUTH19101    PUTH  Putha Hiunchuli  2019  Spring    2019-05-09   
10362     RATC19101    RATC      Ratna Chuli  2019  Spring    2019-04-11   
10363     SANK19101    SANK     Sano Kailash  2019  Spring    2019-04-05   

      highpoint_date termination_date  \
0         1960-05-17              NaN   
1         1969-10-22       1969-10-26   
2         1973-05-06              NaN   
3         1978-10-02       1978-10-05   
4         1979-10-18       1979-10-20   
...              ...              ...   
10359     2019-05-01       2019-05-27   
10360     2019-05-01       2019-05-25   
10361     2019-05-16       2019-05-17   
10362     2019-04-19       2019-05-03   
10363     2019-04-28       2019-04-29   

                                      termination_reason  highpoint_metres  \
0                                    Success (main peak)            7937.0   
1                                    Success (main peak)            7937.0   
2                                    Success (main peak)            7937.0   
3                       Bad weather (storms, high winds)            7000.0   
4                       Bad weather (storms, high winds)            7160.0   
...                                                  ...               ...   
10359                                Success (main peak)            7138.0   
10360                                Success (main peak)            7138.0   
10361  Bad conditions (deep snow, avalanching, fallin...            6350.0   
10362            Lack (or loss) of supplies or equipment            6600.0   
10363                                Success (main peak)            6452.0   

       members  member_deaths  hired_staff  hired_staff_deaths  oxygen_used  \
0           10              0            9                   0         True   
1           10              0            0                   0        False   
2            6              0            8                   0        False   
3            2              0            0                   0        False   
4            3              0            0                   0        False   
...        ...            ...          ...                 ...          ...   
10359        8              0            4                   0        False   
10360        1              0            1                   0        False   
10361        5              0            2                   0        False   
10362        1              0            0                   0        False   
10363        2              0            1                   0        False   

            trekking_agency  
0                       NaN  
1                       NaN  
2                       NaN  
3                       NaN  
4                       NaN  
...                     ...  
10359   Mountain Experience  
10360        Asian Trekking  
10361      Iceland Trekking  
10362     Pioneer Adventure  
10363  Langtang Ri Trekking  

[10364 rows x 16 columns]

In [21]:
df_1=pd.merge(members,expeditions,how="right",on="expedition_id")

In [22]:
df_2=pd.merge(peaks,df_1,how="right",right_on="expedition_id",left_on="first_ascent_expedition_id")

In [23]:
df_2

peak_id     peak_name peak_alternative_name  height_metres  \
0        ANN2  Annapurna II                   NaN         7937.0   
1        ANN2  Annapurna II                   NaN         7937.0   
2        ANN2  Annapurna II                   NaN         7937.0   
3        ANN2  Annapurna II                   NaN         7937.0   
4        ANN2  Annapurna II                   NaN         7937.0   
...       ...           ...                   ...            ...   
76546     NaN           NaN                   NaN            NaN   
76547     NaN           NaN                   NaN            NaN   
76548    SANK  Sano Kailash                   NaN         6452.0   
76549    SANK  Sano Kailash                   NaN         6452.0   
76550    SANK  Sano Kailash                   NaN         6452.0   

      climbing_status  first_ascent_year first_ascent_country  \
0             Climbed             1960.0            UK, Nepal   
1             Climbed             1960.0            UK, Nepal   
2             Climbed             1960.0            UK, Nepal   
3             Climbed             1960.0            UK, Nepal   
4             Climbed             1960.0            UK, Nepal   
...               ...                ...                  ...   
76546             NaN                NaN                  NaN   
76547             NaN                NaN                  NaN   
76548         Climbed             2019.0       Austria, Nepal   
76549         Climbed             2019.0       Austria, Nepal   
76550         Climbed             2019.0       Austria, Nepal   

      first_ascent_expedition_id expedition_id     member_id  ...  \
0                      ANN260101     ANN260101  ANN260101-01  ...   
1                      ANN260101     ANN260101  ANN260101-04  ...   
2                      ANN260101     ANN260101  ANN260101-03  ...   
3                      ANN260101     ANN260101  ANN260101-10  ...   
4                      ANN260101     ANN260101  ANN260101-02  ...   
...                          ...           ...           ...  ...   
76546                        NaN     PUTH19101  PUTH19101-07  ...   
76547                        NaN     RATC19101  RATC19101-01  ...   
76548                  SANK19101     SANK19101  SANK19101-01  ...   
76549                  SANK19101     SANK19101  SANK19101-02  ...   
76550                  SANK19101     SANK19101  SANK19101-03  ...   

      highpoint_date termination_date  \
0         1960-05-17              NaN   
1         1960-05-17              NaN   
2         1960-05-17              NaN   
3         1960-05-17              NaN   
4         1960-05-17              NaN   
...              ...              ...   
76546     2019-05-16       2019-05-17   
76547     2019-04-19       2019-05-03   
76548     2019-04-28       2019-04-29   
76549     2019-04-28       2019-04-29   
76550     2019-04-28       2019-04-29   

                                      termination_reason highpoint_metres_y  \
0                                    Success (main peak)             7937.0   
1                                    Success (main peak)             7937.0   
2                                    Success (main peak)             7937.0   
3                                    Success (main peak)             7937.0   
4                                    Success (main peak)             7937.0   
...                                                  ...                ...   
76546  Bad conditions (deep snow, avalanching, fallin...             6350.0   
76547            Lack (or loss) of supplies or equipment             6600.0   
76548                                Success (main peak)             6452.0   
76549                                Success (main peak)             6452.0   
76550                                Success (main peak)             6452.0   

      members  member_deaths hired_staff hired_staff_deaths oxygen_used_y  \
0          10              0           9                  0 

In [24]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 76551 entries, 0 to 76550
Data columns (total 44 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   peak_id                     2424 non-null   object 
 1   peak_name                   2424 non-null   object 
 2   peak_alternative_name       1273 non-null   object 
 3   height_metres               2424 non-null   float64
 4   climbing_status             2424 non-null   object 
 5   first_ascent_year           2424 non-null   float64
 6   first_ascent_country        2424 non-null   object 
 7   first_ascent_expedition_id  2424 non-null   object 
 8   expedition_id               76551 non-null  object 
 9   member_id                   76538 non-null  object 
 10  peak_id_x                   76538 non-null  object 
 11  peak_name_x                 76523 non-null  object 
 12  year_x                      76538 non-null  float64
 13  season_x                    765

In [26]:
df_2[df_2["sex"]=="F"][["first_ascent_year","sex","member_id","peak_name","height_metres"]].sort_values(by="first_ascent_year")

first_ascent_year sex     member_id         peak_name  height_metres
33272             1949.0   F  DZAN49101-03            Dzanye         6581.0
28221             1949.0   F  TENG49101-03          Tengkoma         6215.0
25481             1955.0   F  LEOE55101-01  Leonpo Gang East         6733.0
25482             1955.0   F  LEOE55101-03  Leonpo Gang East         6733.0
25485             1955.0   F  LEOE55101-02  Leonpo Gang East         6733.0
...                  ...  ..           ...               ...            ...
76495                NaN   F  MAKA19105-01               NaN            NaN
76496                NaN   F  MAKA19105-02               NaN            NaN
76499                NaN   F  MAKA19105-05               NaN            NaN
76505                NaN   F  MAKA19106-02               NaN            NaN
76508                NaN   F  MAKA19107-01               NaN            NaN

[7045 rows x 5 columns]

Here we can see the first record of a woman summiting a peak are DZAN49101-03 and TENG49101-03, the names of the peaks are Dzanye and Tengkoma, and the heights for the peaks are 6581 and 6215. 

b) Provide a single cross-tab table of the sex of an expedition participant and - the proportion of success in summitting a main peak or sub-peak
- the proportion of participants using oxygen support - the proportion of participants who died.

That is, sex should be displayed as the rows of the output and the three requested proportions as its columns.

In [73]:
f=df_2.groupby("sex")
ff=f[["success","oxygen_used_x","died"]].sum()
ff=pd.DataFrame(ff)
ff

success  oxygen_used_x  died
sex                              
F       2306           1521    50
M      26904          16714  1056

In [74]:
ff[0:1]=ff[0:1]/len(df_2[df_2["sex"]=="F"])
ff[1:2]=ff[1:2]/len(df_2[df_2["sex"]=="M"])

In [75]:
ff

success  oxygen_used_x      died
sex                                   
F    0.327324       0.215898  0.007097
M    0.387158       0.240520  0.015196

4. Accidents

a) Aggregate the data by decade and count the number of expeditions, the total number of expedition members, and the average rates of death for members and hired staff. Briefly describe what you found.

In [119]:
exp=expeditions.copy()

In [125]:
exp["decade"]=exp["year"]//10*10

In [129]:
exp[exp["decade"]==1920]

expedition_id peak_id      peak_name  year  season basecamp_date  \
2565     KANG29101    KANG  Kangchenjunga  1929  Spring    1929-05-21   
2754     EVER21101    EVER        Everest  1921  Spring    1921-06-26   
2755     EVER22101    EVER        Everest  1922  Spring    1922-05-01   
2756     EVER24101    EVER        Everest  1924  Spring    1924-04-28   
2859     KABD25101    KABD     Kabru Dome  1925  Spring           NaN   
2865     KABN20401    KABN    Kabru North  1920  Winter           NaN   
2884     KANG20301    KANG  Kangchenjunga  1920  Autumn           NaN   
2885     KANG25101    KANG  Kangchenjunga  1925  Spring           NaN   
2888     KANG29201    KANG  Kangchenjunga  1929  Summer    1929-08-10   
3090     TLNG20301    TLNG         Talung  1920  Autumn           NaN   

     highpoint_date termination_date                      termination_reason  \
2565     1929-05-27              NaN      Accident (death or serious injury)   
2754     1921-09-24       1921-09-30                   Did not attempt climb   
2755     1922-05-27       1922-06-14  Illness, AMS, exhaustion, or frostbite   
2756     1924-06-08       1924-06-15      Accident (death or serious injury)   
2859            NaN              NaN                                   Other   
2865            NaN              NaN                                   Other   
2884            NaN              NaN                                   Other   
2885            NaN              NaN                   Did not attempt climb   
2888     1929-10-03       1929-10-18        Bad weather (storms, high winds)   
3090            NaN              NaN                                   Other   

      highpoint_metres  members  member_deaths  hired_staff  \
2565            6000.0        1              1            3   
2754            7000.0        9              1           21   
2755            8320.0       13              0           30   
2756            8600.0       11              2           40   
2859            6100.0        1              0            0   
2865            6400.0        1              0            0   
2884            6400.0        2              0            0   
2885               NaN        1              0            0   
2888            7400.0       10              0           25   
3090            5800.0        2              0            0   

      hired_staff_deaths  oxygen_used trekking_agency  decade  
2565                   0        False             NaN    1920  
2754                   1        False             NaN    1920  
2755                   7         True             NaN    1920  
2756                   2         True             NaN    1920  
2859                   0        False             NaN    1920  
2865                   0        False             NaN    1920  
2884                   0        False             NaN    1920  
2885                   0        False             NaN    1920  
2888                   0        False             NaN    1920  
3090                   0        False             NaN    1920

In [159]:
t_0=exp.groupby("decade")["members"].sum()
t_1=exp.groupby("decade")["expedition_id"].count()
t_2=exp.groupby("decade")["member_deaths"].sum()
t_3=exp.groupby("decade")["hired_staff_deaths"].sum()
t_4=exp.groupby("decade")["hired_staff"].sum()

In [160]:
t_2=t_2/t_0
t_3=t_3/t_4
t_2=pd.DataFrame(t_2)
t_3=pd.DataFrame(t_3)
t_2

0
decade          
1900    0.111111
1910    0.000000
1920    0.078431
1930    0.013158
1940    0.000000
1950    0.018692
1960    0.022245
1970    0.027667
1980    0.019954
1990    0.013027
2000    0.008182
2010    0.008694

In [161]:
t=pd.merge(t_0,t_1,on="decade")
tt=pd.merge(t_2,t_3,on="decade")
t=pd.merge(t,tt,on="decade")

In [168]:
t=t.rename(columns={"0_x":"member death ratio"})
t=t.rename(columns={"0_y":"hired staff death ratio"})

In [169]:
t

members  expedition_id  member death ratio  hired staff death ratio
decade                                                                     
1900          9              4            0.111111                 1.000000
1910          3              3            0.000000                      NaN
1920         51             10            0.078431                 0.084034
1930        152             34            0.013158                 0.014652
1940         53             15            0.000000                 0.000000
1950        856            168            0.018692                 0.011968
1960        989            161            0.022245                 0.010363
1970       3506            361            0.027667                 0.026557
1980      10023           1187            0.019954                 0.015380
1990      12512           1703            0.013027                 0.011794
2000      16867           3213            0.008182                 0.007660
2010      16679           3505            0.008694                 0.007229

The death ratio for members and hired staff decreases as time goes by and the number of members and expeditions increases gradually. 

b) Calculate the length_of_expedition as the time between the basecamp_date and the termination_date in days. Now, standardize these length_of_expedition values to z-values grouped by each peak (that is you need to standardize with the mean and standard deviation of expeditions for the same peak only). Are longer expeditions more or less likely to be associated with death? (no statistical test necessary)

In [177]:
from datetime import date
exp[["basecamp_date","termination_date"]] = exp[["basecamp_date","termination_date"]].apply(pd.to_datetime) 
exp["length_of_expedition"]=(exp["termination_date"]-exp["basecamp_date"]).dt.days

In [190]:
s=exp["length_of_expedition"].std()

In [191]:
m=exp["length_of_expedition"].mean()

In [192]:
exp["z_value"]=(exp["length_of_expedition"]-m)/s

In [200]:
l=pd.DataFrame(exp.groupby("peak_name")["z_value"].mean())

In [202]:
l

z_value
peak_name              
Aichyn              NaN
Ama Dablam    -0.914590
Amotsang      -1.302238
Amphu Gyabjen -0.198980
Amphu I       -0.750609
...                 ...
Yala Chuli          NaN
Yalung Kang    0.854129
Yanme Kang    -1.334687
Yansa Tsenji  -0.783058
Yaupa         -0.404489

[390 rows x 1 columns]

In [205]:
b=exp.groupby("peak_name")["length_of_expedition"].mean()

In [207]:
v=exp.groupby("peak_name")["member_deaths"].sum()

In [210]:
pd.merge(b,v,on="peak_name")


length_of_expedition  member_deaths
peak_name                                         
Aichyn                          NaN              0
Ama Dablam                12.473239             26
Amotsang                   6.500000              0
Amphu Gyabjen             23.500000              0
Amphu I                   15.000000              0
...                             ...            ...
Yala Chuli                      NaN              0
Yalung Kang               39.727273              6
Yanme Kang                 6.000000              0
Yansa Tsenji              14.500000              0
Yaupa                     20.333333              0

[390 rows x 2 columns]

From the data we got, it seem longer expeditions may lead higher death ratio.